### DISCUSSION TITLE
Dev Proxy to external API

### DISCUSSION THREAD

**Discussion by cclloyd** [vercel/next.js#11508] [Help] [answered]
I've been googling for a while now, trying to figure out how to proxy certain URLs to a separate API, not hosted by Next.js.

My current attempt was to edit the webpack config, because that's wat some issues in this repo said to do.  So I had the following in my `next.config.js`

    const { PHASE_DEVELOPMENT_SERVER } = require('next/constants');
  
    module.exports = (phase, { defaultConfig }) => {
        if (phase === PHASE_DEVELOPMENT_SERVER) {
            return {
                /* development only config options here */
                webpack: (webpackConfig) => {
                    const newConfig = { ...webpackConfig };
                    newConfig.devServer = {
                        ...webpackConfig.devServer,
                        proxy: [{
                            context: ['/api', '/admin', '/staticfiles'],
                            target: 'http://localhost:8000',
                        }]
                    };
                    return newConfig;
                },
                cdn: false,
            }
        }
    }

But that doesn't work.  When I go to `/api/` on my page, it returns the 404 route.  

So what would be the best way to proxy `/api` to `http://localhost:8000` (no rewrite) using the next dev server (`npm run dev` where dev is `next dev`).

**Comment by timneutkens** [comment, marked as answer]
A new feature will be introduced soon to allow for this. You can use the experimental option currently:

```js
// next.config.js
module.exports = {
  experimental: {
    rewrites: () => {
      return [
        source: '/api/:path*',
        destination: 'http://localhost:8000/:path*'
      ]
    }
  }
}
```

**Reply by cclloyd** [reply]
That seems to work but not with a trailing slash.  If it ends in `/` then it shows the 404.  What do I change to allow that?  (Since all my API endpoints require trailing slash)

Edit: just had to add a trailing slash and it worked (since it will expect a trailing slash anyway.  Cannot confirm if it would differentiate fully between trailing slash or not)

My exports, in case anyone is interested:

    module.exports = {
        experimental: {
            rewrites: () => {
                return [
                    {source: '/api/:path*/', destination: 'http://localhost:8000/api/:path*/'},
                    {source: '/staticfiles/:path*', destination: 'http://localhost:8000/staticfiles/:path*'},
                    {source: '/admin/:path*/', destination: 'http://localhost:8000/admin/:path*/'},
                    {source: '/media/:path*', destination: 'http://localhost:8000/media/:path*'},
                    {source: '/api-auth/:path*/', destination: 'http://localhost:8000/api-auth/:path*/'},
                ]
            }
        }
    }

**Reply by ishrak-imam** [reply]
Which version of nextJS have this experimental feature?

**Reply by TheKalpit** [reply]
This is great!
Is this supposed to work on server-side (eg. when an API is called on the server-side)?
For me, it's working fine on client-side, but not on server.
